stdlib/dynarray.mli

```diff
@@ -0,0 +1,622 @@
+
+    [i] must be a valid index. [set] does not add new elements to the
+    array -- see {!add_last} to add an element.
+
+    @raise Invalid_argument if the index is invalid. *)
+
+val length : 'a t -> int
+(** [length a] is the number of elements in the array. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], a new array
+    containing the same elements as [a]. *)
+
+
+(** {1:adding Adding elements}
+
+    Note: all operations adding elements raise [Invalid_argument] if the
+    length needs to grow beyond {!Sys.max_array_length}. *)
+
+val add_last : 'a t -> 'a -> unit
+(** [add_last a x] adds the element [x] at the end of the array [a]. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** [append_array a b] adds all elements of [b] at the end of [a],
+    in the order they appear in [b].
+
+    For example:
+    {[
+      let a = Dynarray.of_list [1;2] in
+      Dynarray.append_array a [|3; 4|];
+      assert (Dynarray.to_list a = [1; 2; 3; 4])
+    ]}
+*)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append_array} but with a list. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] is like [append_array a b],
+    but [b] is itself a dynamic array instead of a fixed-size array.
+
+    Warning: [append a a] is a programming error because it iterates
+    on [a] and adds elements to it at the same time -- see the
+    {{!sec:iteration} Iteration} section below. It fails with
+    [Invalid_argument].
+    If you really want to append a copy of [a] to itself, you can use
+    [Dynarray.append_array a (Dynarray.to_array a)] which copies [a]
+    into a temporary array.
+*)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append_array} but with a sequence.
+
+    Warning: [append_seq a (to_seq a)] simultaneously traverses [a]
+    and adds element to it; the ordering of those operations is
+    unspecified, and may result in an infinite loop -- the new
+    elements may in turn be produced by [to_seq a] and get added again
+    and again.
+*)
+
+val append_iter :
+  'a t ->
+  (('a -> unit) -> 'iter -> unit) ->
+  'iter -> unit
+(** [append_iter a iter x] adds each element of [x] to the end of [a].
+    This is [iter (add_last a) x].
+
+    For example, [append_iter a List.iter [1;2;3]] would add elements
+    [1], [2], and then [3] at the end of [a].
+    [append_iter a Queue.iter q] adds elements from the queue [q]. *)
+
+
+(** {1:removing Removing elements} *)
+
+val pop_last_opt : 'a t -> 'a option
+(** [pop_last_opt a] removes and returns the last element of [a],
+    or [None] if the array is empty. *)
+
+val pop_last : 'a t -> 'a
+(** [pop_last a] removes and returns the last element of [a].
+
+    @raise Not_found on an empty array. *)
+
+val remove_last : 'a t -> unit
+(** [remove_last a] removes the last element of [a], if any.
+    It does nothing if [a] is empty. *)
```

Comment:
I am having second thoughts about the name `remove_last`. I was writing some code and reaching for this function, and I called it `drop_last`. In my mind a `remove` function would also return the removed value, while `drop` highlights that the value is discarded.

The name `remove` comes from @c-cube and some version of it (but not this exact function) is present in Containers:

https://github.com/c-cube/ocaml-containers/blob/ba516e81af89d1cb5e3ae9456fe558e279254ac3/src/core/CCVector.mli#L266-L278

Any thoughts on my proposal to rename to `drop_last`? (@c-cube, @dbuenzli)

(I see that containers has a `top` function that returns the last element, and I wonder if it would make sense to have `get_last{,_opt}` functions as well. )

Reply:
> Any thoughts on my proposal to rename to `drop_last`? (@c-cube, @dbuenzli)

Needless irregularity in my opinion. We have `{Set,Map,Hashtbl}.remove` and `List.remove_assoc`.  The only place where we have `drop` is in `Seq` (an entirely different thing) and `Stack` (the same thing, but again a Dynarray, is not a stack in general). I would just keep `remove_last` in peace. 

> In my mind a `remove` function would also return the removed value, while `drop` highlights that the value is discarded.

None of the functions I mentioned above do that in particular. If anything they return the updated container for immutable ones.

> (I see that containers has a `top` function that returns the last element, and I wonder if it would make sense to have `get_last{,_opt}` functions as well. )

I had these kind of functions in `Pvec` [here](https://github.com/dbuenzli/pvec/blob/383dcc1429b775cf33c8c8a431a9607e6a2646fb/src/pvec.mli#L144-L145) and [there](https://github.com/dbuenzli/pvec/blob/383dcc1429b775cf33c8c8a431a9607e6a2646fb/src/pvec.mli#L144-L145). I would simply have `get_last` raising `Invalid_argument` on the empty array and `last` returning the option.

Reply:
I will leave `remove_last` in piece, thank you both for the quick feedback.